File: packages/server/src/services/smart-pilot.ts
import type { Session, DevelopmentTask, Guidance, CouncilDecision } from '@opencode-autopilot/shared';
import { sessionManager } from './session-manager.js';
import { council } from './council.js';
import { wsManager } from './ws-manager.js';
import { loadConfig } from './config.js';
import { autoContinueHooks } from './hooks.js';

interface TaskCheckpoint {
  sessionId: string;
  lastCheckedAt: number;
  lastTaskId?: string;
  pendingDebate: boolean;
}

interface SmartPilotConfig {
  enabled: boolean;
  pollIntervalMs: number;
  autoApproveThreshold: number;
  requireUnanimous: boolean;
  maxAutoApprovals: number;
}

class SmartPilot {
  private config: SmartPilotConfig;
  private checkpoints: Map<string, TaskCheckpoint> = new Map();
  private pollTimer: ReturnType<typeof setInterval> | null = null;
  private autoApprovalCount: Map<string, number> = new Map();

  constructor() {
    const appConfig = loadConfig();
    this.config = {
      enabled: appConfig.council.smartPilot || false,
      pollIntervalMs: 3000,
      autoApproveThreshold: 0.8,
      requireUnanimous: false,
      maxAutoApprovals: 10,
    };
  }

  setEnabled(enabled: boolean): void {
    this.config.enabled = enabled;
    if (enabled) {
      this.start();
    } else {
      this.stop();
    }
  }

  setAutoApproveThreshold(threshold: number): void {
    this.config.autoApproveThreshold = Math.max(0.5, Math.min(1.0, threshold));
  }

  setRequireUnanimous(require: boolean): void {
    this.config.requireUnanimous = require;
  }

  setMaxAutoApprovals(max: number): void {
    this.config.maxAutoApprovals = Math.max(1, max);
  }

  isEnabled(): boolean {
    return this.config.enabled;
  }

  getConfig(): SmartPilotConfig {
    return { ...this.config };
  }

  async triggerTask(sessionId: string, task: DevelopmentTask): Promise<void> {
    const session = sessionManager.getSession(sessionId);
    if (!session) throw new Error(`Session ${sessionId} not found`);

    // Process immediately without polling
    await this.runDebateAndRespond(session, task);
  }

  start(): void {
    if (this.pollTimer) return;
    if (!this.config.enabled) return;

    console.log('[SmartPilot] Starting auto-pilot mode');
    
    this.pollTimer = setInterval(() => {
      this.checkAllSessions();
    }, this.config.pollIntervalMs);
  }

  stop(): void {
    if (this.pollTimer) {
      clearInterval(this.pollTimer);
      this.pollTimer = null;
      console.log('[SmartPilot] Stopped');
    }
  }

  private async checkAllSessions(): Promise<void> {
    const activeSessions = sessionManager.getActiveSessions();
    
    for (const session of activeSessions) {
      await this.checkSession(session);
    }
  }

  private async checkSession(session: Session): Promise<void> {
    let checkpoint = this.checkpoints.get(session.id);
    
    if (!checkpoint) {
      checkpoint = {
        sessionId: session.id,
        lastCheckedAt: Date.now(),
        pendingDebate: false,
      };
      this.checkpoints.set(session.id, checkpoint);
    }

    if (checkpoint.pendingDebate) return;

    const task = await this.fetchCurrentTask(session);
    if (!task) return;

    if (task.id === checkpoint.lastTaskId) return;

    checkpoint.lastTaskId = task.id;
    checkpoint.pendingDebate = true;

    try {
      await this.runDebateAndRespond(session, task);
    } finally {
      checkpoint.pendingDebate = false;
      checkpoint.lastCheckedAt = Date.now();
    }
  }

  private async fetchCurrentTask(session: Session): Promise<DevelopmentTask | null> {
    const managed = this.getManagedSession(session.id);
    if (!managed) return null;

    try {
      const res = await fetch(`http://localhost:${managed.port}/current-task`);
      if (!res.ok) return null;
      return await res.json();
    } catch {
      return null;
    }
  }

  private getManagedSession(sessionId: string): { port: number } | undefined {
    const allSessions = sessionManager.getAllSessions();
    const session = allSessions.find(s => s.id === sessionId);
    if (!session) return undefined;

    const basePort = loadConfig().sessions.basePort;
    const idx = allSessions.indexOf(session);
    return { port: basePort + idx };
  }

  private async runDebateAndRespond(session: Session, task: DevelopmentTask): Promise<void> {
    console.log(`[SmartPilot] New task detected in ${session.id}: ${task.description}`);
    
    wsManager.notifyLog(session.id, {
      timestamp: Date.now(),
      level: 'info',
      message: `[SmartPilot] Initiating council debate for: ${task.description}`,
      source: 'smart-pilot',
    });

    const preDebateResult = await autoContinueHooks.execute({
      phase: 'pre-debate',
      session,
      task,
    });
    
    if (!preDebateResult.continue) {
      wsManager.notifyLog(session.id, {
        timestamp: Date.now(),
        level: 'warn',
        message: `[SmartPilot] Debate blocked by hook: ${preDebateResult.reason || 'No reason provided'}`,
        source: 'smart-pilot',
      });
      return;
    }

    let decision = await council.debate(task);

    const postDebateResult = await autoContinueHooks.execute({
      phase: 'post-debate',
      session,
      task,
      decision,
    });
    
    if (!postDebateResult.continue) {
      wsManager.notifyLog(session.id, {
        timestamp: Date.now(),
        level: 'warn',
        message: `[SmartPilot] Post-debate hook stopped flow: ${postDebateResult.reason || 'No reason provided'}`,
        source: 'smart-pilot',
      });
      return;
    }
    
    if (postDebateResult.modifiedDecision) {
      decision = postDebateResult.modifiedDecision;
    }

    wsManager.notifyCouncilDecision(session.id, decision);

    let guidance = this.decisionToGuidance(decision, session.id);

    if (guidance.approved) {
      const count = (this.autoApprovalCount.get(session.id) || 0) + 1;
      this.autoApprovalCount.set(session.id, count);

      if (count > this.config.maxAutoApprovals) {
        console.log(`[SmartPilot] Max auto-approvals (${this.config.maxAutoApprovals}) reached for ${session.id}`);
        guidance.approved = false;
        guidance.feedback = `Auto-approval limit reached (${this.config.maxAutoApprovals}). Manual review required.`;
      }
    }

    const preGuidanceResult = await autoContinueHooks.execute({
      phase: 'pre-guidance',
      session,
      task,
      decision,
      guidance,
    });
    
    if (!preGuidanceResult.continue) {
      wsManager.notifyLog(session.id, {
        timestamp: Date.now(),
        level: 'warn',
        message: `[SmartPilot] Guidance blocked by hook: ${preGuidanceResult.reason || 'No reason provided'}`,
        source: 'smart-pilot',
      });
      return;
    }
    
    if (preGuidanceResult.modifiedGuidance) {
      guidance = preGuidanceResult.modifiedGuidance;
    }

    wsManager.notifyLog(session.id, {
      timestamp: Date.now(),
      level: guidance.approved ? 'info' : 'warn',
      message: `[SmartPilot] Decision: ${guidance.approved ? 'AUTO-APPROVED' : 'REQUIRES REVIEW'} (consensus: ${(decision.weightedConsensus || decision.consensus * 100).toFixed(1)}%)`,
      source: 'smart-pilot',
    });

    try {
      await sessionManager.sendGuidance(session.id, guidance);
      
      await autoContinueHooks.execute({
        phase: 'post-guidance',
        session,
        task,
        decision,
        guidance,
      });
    } catch (err) {
      console.error(`[SmartPilot] Failed to send guidance to ${session.id}:`, err);
      
      await autoContinueHooks.execute({
        phase: 'on-error',
        session,
        task,
        decision,
        guidance,
        error: err instanceof Error ? err : new Error(String(err)),
      });
    }
  }

  private decisionToGuidance(decision: CouncilDecision, sessionId: string): Guidance {
    const effectiveConsensus = decision.weightedConsensus ?? decision.consensus;
    
    let canAutoApprove = decision.approved && effectiveConsensus >= this.config.autoApproveThreshold;

    if (this.config.requireUnanimous && decision.votes.some(v => !v.approved)) {
      canAutoApprove = false;
    }

    if (decision.dissent && decision.dissent.length > 0) {
      canAutoApprove = false;
    }

    const suggestedNextSteps: string[] = [];
    
    if (canAutoApprove) {
      suggestedNextSteps.push('Continue with implementation');
    } else {
      suggestedNextSteps.push('Review council feedback');
      if (decision.dissent && decision.dissent.length > 0) {
        suggestedNextSteps.push('Address dissenting concerns');
      }
      suggestedNextSteps.push('Request manual approval if appropriate');
    }

    return {
      approved: canAutoApprove,
      feedback: decision.reasoning,
      suggestedNextSteps,
    };
  }

  resetApprovalCount(sessionId: string): void {
    this.autoApprovalCount.delete(sessionId);
  }

  resetAllApprovalCounts(): void {
    this.autoApprovalCount.clear();
  }

  cleanup(): void {
    this.stop();
    this.checkpoints.clear();
    this.autoApprovalCount.clear();
  }
}

export const smartPilot = new SmartPilot();

File: go-port/pkg/server/services/smartpilot/smart_pilot.go
package smartpilot

import (
	"math"
	"sync"
	"time"

	"borg-orchestrator/pkg/server/services/config"
	"borg-orchestrator/pkg/shared"
)

type TaskCheckpoint struct {
	SessionID     string
	LastCheckedAt int64
	LastTaskID    *string
	PendingDebate bool
}

type SmartPilotConfig struct {
	Enabled              bool
	PollIntervalMs       int
	AutoApproveThreshold float64
	RequireUnanimous     bool
	MaxAutoApprovals     int
}

type SmartPilot struct {
	mu                sync.RWMutex
	config            SmartPilotConfig
	checkpoints       map[string]*TaskCheckpoint
	pollTimer         *time.Ticker
	stopChan          chan struct{}
	autoApprovalCount map[string]int
	activePlans       map[string]shared.TaskPlan
}

var Service = NewSmartPilot()

func NewSmartPilot() *SmartPilot {
	appConfig := config.LoadConfig()
	enabled := false
	if appConfig.Council.SmartPilot != nil {
		enabled = *appConfig.Council.SmartPilot
	}

	return &SmartPilot{
		config: SmartPilotConfig{
			Enabled:              enabled,
			PollIntervalMs:       3000,
			AutoApproveThreshold: 0.8,
			RequireUnanimous:     false,
			MaxAutoApprovals:     10,
		},
		checkpoints:       make(map[string]*TaskCheckpoint),
		autoApprovalCount: make(map[string]int),
		activePlans:       make(map[string]shared.TaskPlan),
	}
}

func (s *SmartPilot) SetEnabled(enabled bool) {
	s.mu.Lock()
	s.config.Enabled = enabled
	s.mu.Unlock()

	if enabled {
		s.Start()
	} else {
		s.Stop()
	}
}

func (s *SmartPilot) SetAutoApproveThreshold(threshold float64) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.config.AutoApproveThreshold = math.Max(0.5, math.Min(1.0, threshold))
}

func (s *SmartPilot) SetRequireUnanimous(require bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.config.RequireUnanimous = require
}

func (s *SmartPilot) SetMaxAutoApprovals(max int) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if max < 1 {
		max = 1
	}
	s.config.MaxAutoApprovals = max
}

func (s *SmartPilot) IsEnabled() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.config.Enabled
}

func (s *SmartPilot) GetConfig() SmartPilotConfig {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.config
}

func (s *SmartPilot) ResetApprovalCount(sessionID string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.autoApprovalCount, sessionID)
}

func (s *SmartPilot) ResetAllApprovalCounts() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.autoApprovalCount = make(map[string]int)
}

func (s *SmartPilot) Cleanup() {
	s.Stop()
	s.mu.Lock()
	defer s.mu.Unlock()
	s.checkpoints = make(map[string]*TaskCheckpoint)
	s.autoApprovalCount = make(map[string]int)
	s.activePlans = make(map[string]shared.TaskPlan)
}
